nexus/src/db/datastore.rs

```diff
@@ -282,10 +310,42 @@ impl DataStore {
                     LookupType::ByName(name.as_str().to_owned()),
                 )
             })
+            .map(|o| (authz::FLEET.organization(o.id()), o))
+    }
+
+    /// Look up the id for an organization based on its name
+    ///
+    /// Returns an [`authz::Organization`] (which makes the id available).
+    ///
+    /// This function does no authz checks because it is not possible to know
+    /// just by looking up an Organization's id what privileges are required.
```

Comment:
Not sure I understand this point. Couldn't you check for the read permission? Are you accounting for the possibility that someone could access a sub-resource without having read permission on the organization?

I wonder if that warrants another permission that's like `READ_THRU` that indicates the ability to use the ID to look up sub-resources. Feels like overkill, though. It would correlate one-to-one with having an actual READ/etc permission on a sub-resource, so I suppose in some sense it would be an optimization.

Reply:
Yeah, I don't think we want to give you full read permissions on the Organization just because you have some permission on something inside the Organization.

We could create a `read_thru` like you suggest.  That permission would be granted implicitly by having any role on anything inside the Organization.  To know if you have that, we'd have to either (a) load all the rules that you have on all objects, and for each one, check if it's inside that Organization [which might involve several hops -- from Instance to Project to Organization, for example].  Or (b) load all the resources in the Organization and check for each one whether you have roles on it.  Both of these are potentially extremely expensive.  It's worse: if you're looking up an Instance, have to do all that for the Organization, then do it all again for the Project.  We could denormalize the representation so that we could more quickly know what permissions people have, but that would make other operations much more expensive (e.g., moving a Project from one Organization to another).

I think _not_ creating a permission for this more closely resembles the user experience, too: users get access to, say, a Project.  That doesn't mean they can do anything with the Organization aside from seeing that it exists because it happens to be part of the Project's path.  (On the other hand, maybe every user implicitly has some read access to their own parent Organization, and that lets them see some basic stuff.  But if someone were to share a specific Project in another Organization with you, I think you should still be able to look it up without somebody giving you some extra other permission on the Organization.)

Reply:
That's all fine, seems like the optimal middle ground. We'll see how we can make it harder to accidentally make the noauth thing `pub`, or (maybe riskier since it does not require making it pub) to use it in some other datastore function without the appropriate auth check.

Reply:
Sounds good.  My (vague, hand-wavy) thinking here is to put the noauth functions into a submodule that itself is not exposed.

Another idea I thought of is to have the fields in the db models all private, with an authz-protected interface for getting a similar struct with the same fields "pub" or a trait that provides access.  Either way, we'd do an authz check before giving an object with that data.  This seems tricky and annoying (yet another type called `Organization`?!) but there might be something worthwhile here.